src/sentry/workflow_engine/processors/detector.py

```diff
@@ -55,6 +57,97 @@ def get_detector_by_event(event_data: WorkflowEventData) -> Detector:
     return detector
 
 
+def get_detectors_by_events_bulk(
+    event_list: list[GroupEvent],
+) -> Mapping[str, Detector]:
+    if not event_list:
+        return {}
+
+    # Separate events by whether they have occurrences or not
+    events_with_occurrences: list[tuple[GroupEvent, IssueOccurrence]] = []
+    events_without_occurrences: list[GroupEvent] = []
+
+    for event in event_list:
+        issue_occurrence = event.occurrence
+
+        if issue_occurrence is None:
+            events_without_occurrences.append(event)
+        else:
+            events_with_occurrences.append((event, issue_occurrence))
+
+    result: dict[str, Detector] = {}
+
+    # Fetch detectors for events with occurrences (by detector_id)
+    if events_with_occurrences:
+        detector_ids = []
+        event_to_detector_id: dict[GroupEvent, int] = {}
+
+        for event, occurrence in events_with_occurrences:
+            detector_id = occurrence.evidence_data.get("detector_id")
+            if detector_id is not None:
+                detector_ids.append(detector_id)
+                event_to_detector_id[event] = detector_id
+
+        if detector_ids:
```

Comment:
is there anything we can share with the other case? it feels like we're really just trying to build a list of detector ids in either case.